packages/lodestar/src/chain/opPools/aggregatedAttestationPool.ts

```diff
@@ -0,0 +1,216 @@
+import bls from "@chainsafe/bls";
+import {SLOTS_PER_EPOCH} from "@chainsafe/lodestar-params";
+import {allForks, Slot, ssz, ValidatorIndex} from "@chainsafe/lodestar-types";
+import {CachedBeaconState, phase0} from "@chainsafe/lodestar-beacon-state-transition";
+import {List, toHexString} from "@chainsafe/ssz";
+import {MapDef} from "../../util/map";
+import {pruneBySlot} from "./utils";
+import {InsertOutcome} from "./types";
+
+type DataRootHex = string;
+/**
+ * Maintain a pool of aggregated attestations. Attestations can be retrieved for inclusion in a block
+ * or api. The returned attestations are aggregated to maximise the number of validators that can be
+ * included.
+ * Note that we want to remove attestations with attesters that were included in the chain.
+ */
+export class AggregatedAttestationPool {
+  private readonly attestationGroupByDataHashBySlot = new MapDef<
+    phase0.Slot,
+    Map<DataRootHex, MatchingDataAttestationGroup>
+  >(() => new Map<DataRootHex, MatchingDataAttestationGroup>());
+  // no need to maintain lowestPermissibleSlot since it's checked via gossip validation fn
```

Comment:
We must have this check here too. When our node gets busy promises can be pending for many seconds even minutes. This protection is useful against ourselves

Reply:
added that!